code/messaging/src/main/java/com/adobe/marketing/mobile/messaging/MessagingEventHandler.kt

```diff
@@ -21,25 +21,99 @@ import com.adobe.marketing.mobile.services.Log
  *
  * @property callback the callback to be invoked when the event occurs
  */
-internal open class MessagingEventHandler(private val callback: ContentCardCallback?) {
+internal abstract class MessagingEventHandler(private val callback: ContentCardCallback?) {
     companion object {
         private const val SELF_TAG = "MessagingEventHandler"
+    }
+
+    /**
+     * Method to be implemented by the subclass to perform the template specific logic on events.
+     *
+     * @param event the event to handle
+     * @param propositionId the id of the proposition
+     */
+    abstract fun handleEvent(event: UIEvent<*, *>, propositionId: String)
+
+    /**
+     * Performs the logic common for all templates for display and dismiss events.
+     * Delegates the template specific logic and interact event handling to the subclass by calling [handleEvent].
+     *
+     * @param event the event to handle
+     * @param propositionId the id of the proposition
+     */
+    internal fun onEvent(event: UIEvent<*, *>, propositionId: String) {
+
+        when (event) {
+            is UIEvent.Display -> {
+                Log.trace(
+                    MessagingConstants.LOG_TAG,
+                    SELF_TAG,
+                    "${event.aepUi.getTemplate().getType()} with id $propositionId is displayed"
+                )
+
+                // UIEvent.Display can be called multiple times on configuration changes
+                // We only need to send tracking events for initial composition
+                if (event.aepUi.getState().displayed) {
+                    Log.debug(
+                        MessagingConstants.LOG_TAG,
+                        SELF_TAG,
+                        "UI already displayed, skipping handling display event"
+                    )
+                    return
+                }
+
+                Log.debug(
+                    MessagingConstants.LOG_TAG,
+                    SELF_TAG,
+                    "UI displayed for the first time, sending display tracking event"
+                )
 
-        internal fun track(propositionId: String, interaction: String?, eventType: MessagingEdgeEventType?) {
-            val contentCardSchemaData = ContentCardMapper.instance.getContentCardSchemaData(propositionId)
-            contentCardSchemaData?.track(interaction, eventType)
+                track(propositionId, null, MessagingEdgeEventType.DISPLAY)
+                handleEvent(event, propositionId)
+                callback?.onDisplay(event.aepUi)
```

Comment:
nit: Can we filp the order so that state handling is done before tracking and callbacks.

```
handleEvent(...)
track(..)
callback.onDisplay()
```